File: Sqrt_x.py
'''
Created on 2014.8.16

Leetcode : Decode_Ways

Implement int sqrt(int x).

Compute and return the square root of x.

'''

class Solution:
    # @param x, an integer
    # @return an integer
    def sqrt(self, x):
        if x == 0:
            return 0
        cur = 1;  
        while 1: 
          
            pre = cur;  
            cur = x / (2 * pre) + pre / 2.0;  
            if abs(cur - pre) < 0.00001:
                break
        return int(cur);  
    
test = Solution()
x = 9
print test.sqrt(x)

File: Cplusplus/sqrt_x.cpp
/*
Leetcode : sqrt_x.cpp
Implement int sqrt(int x).

Compute and return the square root of x.
*/
#include <iostream>
#include <climits>
using namespace std;

class Solution {
public:
    int sqrt(int x) {
        unsigned long long left = 1;
        unsigned long long right = x/2;
        while(left <= right)
        {
            unsigned long long mid = (left+right)/2;
            unsigned long long square = mid*mid;
            if(square == x)
                return mid;
            else if(square > x )
            {
                right = mid - 1;
            }
            else
                left = mid + 1;
            //cout<<mid<<endl;
        }
        return left;
    }
};
int main()
{
    int x = 1579205274;
    Solution test = Solution();
    cout<<test.sqrt(x)<<" "<<INT_MAX<<endl;
}
